### ISSUE TITLE
Minesweeper: A crash occurs when the field's size exceeds 32*31 cells

### ISSUE THREAD

**Issue by computerspieler** [SerenityOS/serenity#6283]
Hello everyone,
When I tried to run Minesweeper in "Madwoman" difficulty, the systen froze during a few seconds, while printing the following crash report ( https://pastebin.com/1ihJ8As6 ), and then the task died and the system was back to normal.
And after messing with the field's size, I discovered that Minesweeper would crash when the field's size exceeds 32*31 cells.

**Comment by bcoles** [comment]
```
Minesweeper(26): perror(): anon_create: Too many open files
[#0 Minesweeper(26:26)]: Unrecoverable page fault, write to address V0x00000000
[#0 Minesweeper(26:26)]: Note: Address V0x00000000 looks like a possible nullptr dereference
[Minesweeper(26:26)]: CRASH: CPU #0 Page Fault in ring 3
[#0 Minesweeper(26:26)]: Exception code: 0006 (isr: 0000)
[#0 Minesweeper(26:26)]:     pc=001b:08fb1260 eflags=00010212
[#0 Minesweeper(26:26)]:  stack=0023:01bff400
[#0 Minesweeper(26:26)]:     ds=0023 es=0023 fs=0023 gs=002b
[#0 Minesweeper(26:26)]:    eax=00000000 ebx=2df519f8 ecx=00000828 edx=00000000
[#0 Minesweeper(26:26)]:    ebp=01bff408 esp=c4f91fd8 esi=01c04000 edi=00000000
[#0 Minesweeper(26:26)]:    cr0=80010013 cr2=00000000 cr3=03ec7000 cr4=00140ee4
[#0 Minesweeper(26:26)]: 0x08fb1260  (?)

[#0 Minesweeper(26:26)]: 0xc036b7b0  Kernel::Process::crash(int, unsigned int, bool) +0x3e2
[#0 Minesweeper(26:26)]: 0xc0545efa  illegal_instruction_handler +0x0
[#0 Minesweeper(26:26)]: 0xc0546e2f  page_fault_handler +0xc20
[#0 Minesweeper(26:26)]: 0xc053d30a  page_fault_asm_entry +0x26
```

```
CrashDaemon(15): New coredump file: /tmp/coredump/Minesweeper_26_1618300017
Shell(22): Job entry 'Minesweeper' deleted in 9179 ms
CrashDaemon(15): --- Backtrace for thread #0 (TID 26) ---
CrashDaemon(15): 0x08fb1260: [libc.so] memcpy +0x10 (string.cpp:138)
CrashDaemon(15): 0x2df35840: [libgfx.so] Gfx::PaletteImpl::clone() const +0xa0 (RefPtr.h:189)
CrashDaemon(15): 0x2df35c80: [libgfx.so] Gfx::Palette::set_color(Gfx::ColorRole, Gfx::Color) +0x50 (Atomic.h:266)
CrashDaemon(15): 0x4e6dcf82: [/bin/Minesweeper] Field::reset() +0xcf2 (Atomic.h:266)
CrashDaemon(15): 0x4e6dd7bc: [/bin/Minesweeper] Field::set_field_size(unsigned long, unsigned long, unsigned long) +0x3bc (Widget.h:103)
CrashDaemon(15): 0x4e6e1626: [/bin/Minesweeper] AK::Function<void (GUI::Action&)>::CallableWrapper<main::{lambda(auto:1&)#8}>::call(GUI::Action&) const +0x26 (Function.h:111)
CrashDaemon(15): 0x447eebf6: [libgui.so] GUI::Action::activate(Core::Object*) +0x106 (Atomic.h:266)
CrashDaemon(15): 0x448f2e48: [libgui.so] GUI::WindowServerConnection::handle(Messages::WindowClient::MenuItemActivated const&) +0x58 (WindowServerConnection.cpp:273)
CrashDaemon(15): 0x448f53b0: [libgui.so] .L6144 +0x20 (WindowClientEndpoint.h:2327)
CrashDaemon(15): 0x448f4c68: [libgui.so] IPC::Connection<WindowClientEndpoint, WindowServerEndpoint>::handle_messages() +0xa8 (OwnPtr.h:191)
CrashDaemon(15): 0x448f4d74: [libgui.so] AK::Function<void ()>::CallableWrapper<IPC::Connection<WindowClientEndpoint, WindowServerEndpoint>::Connection(WindowClientEndpoint&, AK::NonnullRefPtr<Core::LocalSocket>)::{lambda()#2}>::call() const +0x64 (Atomic.h:266)
CrashDaemon(15): 0x9a40662c: [libcore.so] Core::Object::dispatch_event(Core::Event&, Core::Object*) +0x7c (Object.h:103)
CrashDaemon(15): 0x9a3efe94: [libcore.so] Core::EventLoop::pump(Core::EventLoop::WaitMode) +0x3e4 (Atomic.h:266)
CrashDaemon(15): 0x9a3f045b: [libcore.so] Core::EventLoop::exec() +0x4b (EventLoop.cpp:390)
CrashDaemon(15): 0x447f1725: [libgui.so] GUI::Application::exec() +0x25 (Application.cpp:120)
CrashDaemon(15): 0x4e6d9f72: [/bin/Minesweeper] main +0xed2 (NonnullRefPtr.h:295)
CrashDaemon(15): 0x4e6da255: [/bin/Minesweeper] _start +0x45 (crt0_shared.cpp:60)
```

**Comment by Granddave** [comment]
I added a debug print that shows the reference counting exceeds some limit.

`Userland/Libraries/LibGfx/Palette.cpp`
```c++
void Palette::set_color(ColorRole role, Color color)
{
    if (m_impl->ref_count() != 1) {
        dbgln("Palette::set_color - m_impl->ref_count(): {}", m_impl->ref_count());
        m_impl = m_impl->clone();
    }
    auto& theme = const_cast<SystemTheme&>(impl().theme());
    theme.color[(int)role] = color.value();
}
```
This occurred when I, as @computerspieler mentioned, changed the game mode to "Madwoman". 

```
...
Minesweeper(26): Palette::set_color - m_impl->ref_count(): 1023
Minesweeper(26): Palette::set_color - m_impl->ref_count(): 1024
Minesweeper(26): perror(): anon_create: Too many open files
...
```

**Comment by boricj** [comment]
It's running out of file descriptors because of egregious Palette cloning. I have a fix inside #6295.
